physical/raft/fsm.go

```diff
@@ -76,6 +85,65 @@ type FSMApplyResponse struct {
 	EntrySlice []*FSMEntry
 }
 
+type logVerificationChunkingShim struct {
+	chunker *raftchunking.ChunkingBatchingFSM
+}
+
+// Apply implements raft.BatchingFSM.
+func (s *logVerificationChunkingShim) Apply(l *raft.Log) interface{} {
+	return s.ApplyBatch([]*raft.Log{l})[0]
+}
+
+// ApplyBatch implements raft.BatchingFSM
+func (s *logVerificationChunkingShim) ApplyBatch(logs []*raft.Log) []interface{} {
+	// This is a hack because raftchunking doesn't play nicely with lower-level
+	// usage of Extensions field like we need for LogStore verification.
+
+	// When we write a verifier log, we write a single byte that consists of the verifierCheckpointOp,
+	// and then we encode the verifier.ExtensionMagicPrefix into the raft log
+	// Extensions field. Both of those together should ensure that verifier
+	// raft logs can never be mistaken for chunked protobufs. See the docs on
+	// verifier.ExtensionMagicPrefix for the reasoning behind the specific value
+	// that was chosen, and how it ensures this property.
+
+	// So here, we need to check for the exact conditions that we encoded when we wrote the
+	// verifier log out. If they match, we're going to insert a dummy raft log. We do this because 1) we
+	// don't want the chunking FSM to blow up on our verifier op that it won't understand and
+	// 2) we need to preserve the length of the incoming slice of raft logs because raft expects
+	// the length of the return value to match 1:1 to the length of the input operations.
+	newBatch := make([]*raft.Log, 0, len(logs))
+
+	for _, l := range logs {
+		if s.isVerifierLog(l) {
+			newBatch = append(newBatch, &raft.Log{})
```

Comment:
Hmmm. I'm not so sure about this. A zero log happens to have type `raft.LogCommand` because that constant has a value of 0.

That means our FSM will attempt to parse the body as protobuf. If it fails the FSM will panic.

I'm guessing since you have tested this out presumably that we are relying on the fact that a nil slice doesn't fail protobuf decode but just decodes a zero-value `LogData`. (though that seems mildly surprising to me at a first glance).

The next potential issue comes from the fact that we then assume that all Logs will have an Index set, however I think we are saved just barely because we have a belt-and-braces safety check to ensure we only use the Index if it's not lower than before so a zero would be ignored...

```
	// Only advance latest pointer if this log has a higher index value than
	// what we have seen in the past.
	var logIndex []byte
	var err error
	latestIndex, _ := f.LatestState()
	lastLog := logs[numLogs-1]
	if latestIndex.Index < lastLog.Index {
```

we then hit this case

```
			switch command := commandRaw.(type) {
			case *LogData:
```

And again are saved only by the fact that our current `ApplyBatch` implementation simply loops over the operations in the `LogData` which will be a nil slice and so will just be a no-op rather than an error.

Raft chunking is safe because it simply checks if it's a LogCommand (which it is) but also sees nil extensions and so just passes it straight to the underlying FSM.

## Conclusion

While this does work today apparently, it feels like we are avoiding a panic only by relying on some subtle implementation details in our FSM in several places.

I think there are a couple of options here:
 1. modify `FSM.ApplyBatch` so it explicitly handles empty logs and treats them as no-ops throughout rather than relying on chance that they don't cause unexpected behavior. At first this seemed simple - just `continue` the loop if there is an empty one, however I think it would end up a bit more involved because of the same reason you added this in the first-place - because we expect to output the same corresponding number of results so just skipping early would break that. You could still explicitly handle empty in each part though.
 2. Maybe simper: Here we could set the raft.Log to be a valid `raft.Log` with the correct index set and with an actual `Data` payload that decodes to an actual FSM transaction with a single `NoOp` command or something.

I'm actually not sure which is better. I don't think it has to be super complex just it feels risky to have so many "near misses" when handling this and rely on them staying that way!

Reply:
My main hesitation with your second option is that I remember adding a `getOp` for the undo logs change being disruptive in the sense that when you had a cluster of nodes on mixed versions, perhaps during an upgrade, the nodes on the older version of Vault that didn't know about `getOp` would see it and panic. We had to write a whole mechanism and polled the versions of every node in the cluster and only enabled undo logs when all nodes were >= 1.13 in order to avoid the panics. I'm looking at the FSM code now though and it's possible we changed it to not return an error when there's an unknown raft op, and thus avoid the panic.

I'll tinker with both of these options and see which one feels better.